gmon CLI

ocervell opened the pull request:
Add gmon CLI - a CLI to query Cloud Monitoring API and ease-up getting / inspecting / deleting metric descriptors, SLOs and Cloud Operations accounts.

jaketf commented on tools/gmon/gmon/clients/utils.py:
```diff
@@ -0,0 +1,47 @@
+from google.protobuf.json_format import MessageToDict
+
+
+def decorate(decorator, methods=[]):
```
I'd personally do something like this

```python3

class DecorateMethods:
    def __init__(self, methods: Set[str]):
        self.methods = methods

    def with(decorator: Callable):
          def inner(cls): 
              for attr in cls.__dict__:
                  attribute = getattr(cls, attr)
              if callable(attribute) and attribute.__name__ in methods:
                  setattr(cls, attr, decorator(attribute))
              return cls

          return inner
```
This way usage reads like english
```python3
@DecorateMethods({'foo', 'bar'}).with(to_json)
```

ocervell replied:
This looks great, but I'm getting into an error when doing this: 
```
@decorate_methods({'get', 'create', 'list', 'delete', 'inspect'}).with(to_json)
                                                                                                      ^
SyntaxError: invalid syntax
```

ocervell replied:
Also tried to rename `with` (Python keyword) to `on`, but doing this fails to decorate the functions:
```
decorator = decorate_methods({'get', 'create', 'list', 'delete',
                              'inspect'}).on(to_json)


@decorator
class MetricsClient:
```

ocervell replied:
How about renaming my existing decorator to `decorate_with`, so it can be used like:
```
@decorate_with(to_json, methods=['get', 'create', 'list', 'delete', 'inspect'])
class MetricsClient
```

jaketf replied:
this makes a lot of sense and is in fact simpler.

jaketf commented on tools/gmon/gmon/clients/utils.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def to_json(func):
+    def inner(*args, **kwargs):
```
use [functools.wraps](https://docs.python.org/3/library/functools.html#functools.wraps) so that stacktraces / docs are still readable with original function name.

jaketf commented on tools/gmon/gmon/clients/utils.py:
```diff
@@ -0,0 +1,47 @@
+"""
```
missing license header

jaketf commented on tools/gmon/gmon/cli.py:
```diff
@@ -0,0 +1,425 @@
+    if not filters:
+        return response
+    dotted = DotMap(response)
+    for k, v in filters.items():
```
these filters can't be pushed server side into SD API request?

ocervell replied:
unfortunately no, the client doesn't allow filtering the response fields (unlike `gcloud`)

ocervell commented on tools/gmon/gmon/cli.py:
```diff
@@ -0,0 +1,425 @@
+    accounts_sub.add_parser('list', help='List Stackdriver Accounts')
+
+    # TODO: Uncomment this when `delete` and `projects.delete`
+    # operations are available.
```
The API is in EAP, so this endpoint is not currently released. Once the API goes alpha / beta I will revisit this.

ocervell commented on tools/gmon/gmon/utils.py:
```diff
@@ -0,0 +1,48 @@
+
+
+def lookup(self, path):
+    dotted_path = path.split(".")
```
Thanks, not sure if I understand the new implementation and I heard using `eval` is unsafe, but it looks nice !

ocervell commented on tools/gmon/gmon/clients/utils.py:
```diff
@@ -0,0 +1,47 @@
+    """
+    if fields is None:
+        return response
+    return {k: v for k, v in response.items() if k in fields}
```
nice !

ocervell commented on tools/gmon/gmon/clients/utils.py:
```diff
@@ -0,0 +1,47 @@
+
+def decorate(decorator, methods=[]):
+    def inner(cls):
+        for attr in cls.__dict__:  # there's propably a better way to do this
```
if it's empty, wouldn't we just not decorate the class instead ?

jaketf replied:
I think I'm being a bit nit picky for the purpose of this tool, but in general we should be careful decorating "all" methods of a class

IMO yes you should specify the methods to be decorated for two reasons:
1) 
> Explicit better than implicit
[PEP20 "Zen of Python"](https://www.python.org/dev/peps/pep-0020/#:~:text=Explicit%20is%20better%20than%20implicit,Flat%20is%20better%20than%20nested.)

2) you may accidentally end up decorating properties (which look like methods) or some methods you don't typically think about to e.g. `__init__` `__eq__`. 

Typically monkey patching is "at your own risk" but It will be very confusing if someone monkey patches a method into your class why their method is behaving differently (e.g. because your decorator will decorate their method).

I think the general principal is decorators should be applied pretty explicitly.

ocervell commented:
@jaketf thanks for your thorough review. I have fixed the few knits :)

ocervell commented:
@jaketf thanks for your thorough review. I have fixed the few knits :)

jaketf commented on tools/gmon/README.md:
```diff
@@ -0,0 +1,141 @@
+here is a cheatsheet:
+
+    # List all metrics in project
+    gmon metrics list -p <PROJECT_ID>
```
bit `gmon` invocations should be formatted w/ backticks

ocervell replied:
done

jaketf commented on tools/gmon/README.md:
```diff
@@ -0,0 +1,141 @@
+# Cloud Monitoring CLI
```
this readme could use an "installation" section.
e.g. how to get this added tool easily and permanently added to your path w/ necessary deps.

ocervell replied:
added

ocervell commented:
Added Pylint, Flake8 checks, a Makefile to run automated CI tests, a cloudbuild.yaml file to build the CLI with Cloud Build, formatted code according to errors, added unit tests.

Coverage is not perfect, but some APIs are difficult to mock (no spec, EAP API):
```
coverage report --rcfile=".coveragerc"
Name                                 Stmts   Miss  Cover
--------------------------------------------------------
gmon/__init__.py                         0      0   100%
gmon/cli.py                            183     21    89%
gmon/clients/__init__.py                 0      0   100%
gmon/clients/account.py                 72     27    62%
gmon/clients/monitoring.py             104     63    39%
gmon/clients/service_monitoring.py      58     24    59%
gmon/clients/utils.py                   30      3    90%
gmon/utils.py                           18      2    89%
--------------------------------------------------------
TOTAL                                  465    140    70%
```

jaketf commented on tools/gmon/tests/unit/test_utils.py:
```diff
@@ -0,0 +1,73 @@
+
+import unittest
+
+# from slo_generator.utils import (get_human_time, get_backend_cls,
```
why are these tests commented out?

ocervell replied:
it was a reliquat from another tool, deleted.

jaketf reviewed (approved):
LGTM
